portfolio/src/main/java/com/google/sps/servlets/ThumbsDownServlet.java

```diff
@@ -31,42 +31,80 @@
 import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 
 @WebServlet("/thumbsdown-data")
 public final class ThumbsDownServlet extends HttpServlet {
 
     @Override
-    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {        
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        UserService userService = UserServiceFactory.getUserService();
+        if (!userService.isUserLoggedIn()) {
+            response.sendRedirect("/contact.html");
+            return;
+        }
+
+        Entity userInfoEntity = getUserInfoEntity();
+        
         // Get comment's id (which was passed as a parameter).
         long id = Long.parseLong(request.getParameter("id"));
 
+        Entity commentEntity = getCommentEntity(id);
+        if (commentEntity == null) {
+            response.setContentType("text/html;");
+            response.getWriter().println("Unable to get comment.");
+            return;
+        }
+
+        if (!isUnlikedComment(userInfoEntity, commentEntity)) {
+            decrementPopularity(commentEntity);
+            addToUnlikedComments(userInfoEntity, commentEntity);
+        } else {
+            incrementPopularity(commentEntity);
+            removeFromUnlikedComments(userInfoEntity, commentEntity);
+        }
+
+
+        response.sendRedirect("/contact.html");
+        return;
+    }
+
+    // Accesses the datastore to get the UserInfo entity. Returns the entity or null if one does not exist.
+    private Entity getCommentEntity(long id) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
         // Using the id, get the comment's key.
         Key commentEntityKey;
         try {
             commentEntityKey = KeyFactory.createKey("Comment", id);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment's key.");
-            return;
+            return null;
         }
 
-        // Instantiate datastore
-        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-
         // Get the comment entity using its key.
         Entity commentEntity;
         try {
             commentEntity = datastore.get(commentEntityKey);
         } catch(Exception e) {
-            response.setContentType("text/html;");
-            response.getWriter().println("Unable to get comment.");
-            return;
+            return null;
         }
 
+        return commentEntity;
+    }
+
+    private void decrementPopularity(Entity commentEntity) {
```

Comment:
You'll need to use a transaction for this method and for any others that modify a database record.

Reply:
Sounds good! I'll address this in a new PR that add transactions if that's okay.